treewide: Deprecate platform aliases

Ericson2314 opened the pull request:
The first commit is succinct and does the depredations. You can just read it, but they are:

1. Left in this PR.
   ```
   system ==> stdenv.buildPlatform.system
   ```
2. Left in this PR.
   ```
   stdenv.system ==> stdenv.buildPlatform.system
   ```
3. Proposed in other open PR.
   ```
   configureFlags = "--foo --bar"; ==> configureFlags = [ "--foo" "--bar" ];
   ```
4. Message improved in other PR
   ```
   isArm ==> isAarch32 # Already done during 18.03
   ```
5. Done in other PR
   ```
   buildPlatform ==> stdenv.buildPlatform
   ```
6. Done in other PR
   ```
   hostPlatform ==> stdenv.hostPlatform
   ```
7. Done in other PR
   ```
   targetPlatform ==> stdenv.targetPlatform
   ```

The rest make Nixpkgs itself comply. If this is accepted, sometime after 18.09 is forked I'll merge another PR to actually remove them.

###### Motivation for this change

See #27069. This makes there be only one non-deprecated way to inspect the build/host/target platform in most cases. `stdenv.is*` is kept, unlike as in in the issue, since it is the most popular and already a shorthand.

The `configureFlags` change is unrelated except I earlier did the work of making nixpkgs obey it. It would be nice to deprecate that too so the work doesn't just bit-rot away.

###### Things done

No hashes should be changed.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
stdenv: Deprecate `stdenv.system`

 - `stdenv.hostPlatform.system` is more explicit. Even if you don't care
 about cross, making clear the intent is good for us cross maintainers
 who might want to make your package work cross-buildable.

 - `system`, wherever it is bound, is rarely the ideal tool for the job.
 Usually it either is too fine-grained or coarse-grained. If one just
 cares about CPU arch or OS, it is bad to inspect both at once with
 `system` because the code will break if a new arch or OS is added. If
 one cares about everything, say for prebuilt binaries, `system` doesn't
 include the ABI, which could lead to broken binaries, or the inability
 to provide both.

dezgeg commented:
Nah, I don't think there's a need to deprecate anything.

LnL7 commented:
I'm probably getting confused by build/host/target again, but shouldn't the source tarball conditions use `targetPlatform`?

Ericson2314 commented:
@dezgeg So straight remove? You've asked that top-level `{build,host,target}Platform` go away.

@LnL7 it should indeed be `hostPlatform`. `targetPlatform` for "source" would just be used for prebuilt standard libraries for a cross compiler.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=949d40bb-ad5f-492c-b7bd-f13f1b66c382)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=347d5f1f-09e3-4abe-af82-e6ac8297c7a0)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=43134788-1ea3-4398-95f9-9636d3c80f5b)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/i6vl5lwlz5jbkg4r6p340dwmj6fha3xq-stdenv-linux
```
</p></details>

Ericson2314 commented:
Actually `isArm` was deprecated in 18.03, so it can be removed immediately.

dezgeg commented:
No, not deprecating of course implies not removing. It appears that toplevel `buildPlatform` etc. have existed since like 17.09, so they cannot be removed without breaking code. What can still be done is to have in-tree uses of Nixpkgs use `stdenv.fooPlatform` consistently.

`stdenv.system` can be made work correctly when cross compiling by setting `passthru.system = hostPlatform.system` in stdenv, so such code is fine.

I don't understand how this `configureFlags` stuff is related at all. No reason to do that either.

> I'm probably getting confused by build/host/target again, but shouldn't the source tarball conditions use targetPlatform?

I've started to think that having `targetPlatform` as a 'global' attribute in the stdenv was a mistake. Among the confusing autotools platform terminology it always sounds the 'correct' one wrt. terms that most people usually use, but in reality it is only ever relevant to things like compilers, assemblers or linkers. Thus it should be an explicit input parameter to the gcc, binutils etc. derivations and not global so people can't use it accidentally.

LnL7 commented on pkgs/development/compilers/rust/bootstrap.nix:
```diff
@@ -16,19 +16,19 @@ let
+    else if stdenv.hostPlatform.system == "x86_64-darwin"
     then "x86_64-apple-darwin"
-    else throw "missing bootstrap url for platform ${stdenv.system}";
+    else throw "missing bootstrap url for platform ${stdenv.hostPlatform.system}";
```
@Ericson2314 Ah I think I know why I was confused about it now, this is an example of a binary tarball, there's no build steps except patching binaries. This and any other cases like it should be using targetPlatform, right?

Ericson2314 replied:
Nope, host platform. @Dezeg is right that "target" sounds great but isn't what you want, and is therefore sort of a trap.

Ericson2314 replied:
just remember target platform is "bad", and then 99% of cases are handled by process of elimination.

Mic92 commented:
Merge conflict originates from removing hurd.

Ericson2314 commented:
@dezgeg as to just removing `targetPlatform`, I am highly sympathetic, but the alternative of referring to `targetPackages.hostPlatform` is just as bad, because we would like each bootstrapping stage to only depend on the previous stage as much as possible. My preferred long shot is just to make all the compilers multi target :D.

Ericson2314 commented:
> I don't understand how this configureFlags stuff is related at all. No reason to do that either.

I edited the OP to address that. The reason is the same as everything else. No reason to support multiple ways of doing things, it's just a maintenance burden and makes packages' source harder to decipher and imitate for people new to nixpkgs.

dezgeg commented:
Because as an end user, once I've written something with a language/library/tool $thing and I have it working, I want it to stay working.

If/when I'm going to upgrade $thing I do it to get bugfixes, new features or to keep getting support from upstream at all. What I don't want to get is to find out I have to go spend time changing my code just to have it working just like it worked before. Especially not if the next day I find some problem with the new version of $thing and can't just downgrade back to the old $thing because now I've changed my code to use the new names the old $thing lacks.

Of course sometimes it's hard or impossible avoid (potential) breakage, e.g. when splitting a package into multiple outputs you really can't do anything sensible to prevent usages like `-I${libfoo}/include` from breaking. But this case is different, it's like, three lines of `inherit buildPlatform hostPlatform targetPlatform;` you need to keep around to not break existing code.

Likewise, when some things like the cross stuff are in the flux, it's more ok to change them. But these ` stdenv.system == "i686-linux"` sort of checks are not new or uncommon, even the oldest release (13.10) has hundreds of those. So by now there's quite a bunch of people who know this interface which has stayed the same for 5 years. Renaming it to something else just for the sake of renaming is just huge collective waste of time for everyone to learn the new stuff.

Ericson2314 commented:
I still like eventual removal, and would happily stgue for that. But first, what about *just* deprecating (as e.g. Java does)?

ElvishJerricco commented:
At least with xyzPlatform, @dezgeg says these have only existed since 17.09. Honestly, that's a fairly short time frame. If we *ever* want to remove them, this is the absolute optimal time to start warning people that it could happen in the future. So the question becomes: Are we to support this API until the end of nixpkgs as we know it? If we expect we'd want them removed at any point in the distant future, we have to start discouraging their use now, or we'll never be able to do it.

IMO (again, only w.r.t. to the xzyPlatform aspect of this PR; sidenote, this PR does too many things), the difference is very minor either way. The breakage should be tiny both because of the newness of the symbols and the relative niche-ness of caring about them right now. But the benefit also isn't huge. However, the cost of having something that's slightly wrong is the kind of thing that can balloon in the future due to unforeseen circumstances, whereas current breakage is more easily understood immediately. So I lean toward adding these warnings for now and removing if and only if it ever makes sense.

Ericson2314 commented:
I give you exhibit https://github.com/nixos/nixpkgs/commit/7ade207f6b75da0fde94cec621ac6d8fa6c3e586 for a rename that never happened, and only got harder.

oxij commented:
I'm :+1: for the rename.

But after similar things of mine were reverted in #44107 (https://github.com/NixOS/nixpkgs/pull/44107#issuecomment-410408564), 4d1332e507fe55103bc5ddd0a384447dc7833e12 and 5c897b4effc4f51b2d6aa5dba158a839b179d964 I now surmise that many people here simply hate the warning messages.

Please enlighten me if there's another explanation. But if the rename + alias doesn't break anything (as it should) what is the argument against having better names, except the warnings? Not to say that this argument does not have a point, as things that have changed recently are more likely to change again soon, so disliking messages that point to the places that you are likely to change multiple times has a point.

Which bring me to the suggestion I was thinking about since those reverts. Can't we adopt two-step rename-deprecation process in Nixpkgs (similarly to how we do that and many similar things in SLNOS)?

I mean,

- commonly in nixpkgs: rename + alias + deprecation warning -> (possibly multiple times) -> eventual deletion of the old attrs;
- commonly in SLNOS: rename + alias -> (possibly multiple times) -> deprecation warning when things got stable -> eventual deletion.

This way we can change our minds about the names multiple times without most users even noticing.

(I passionately want #44107 reverted and the rename plan outlined in https://github.com/NixOS/nixpkgs/pull/44107#issuecomment-410408564 continued, but doing it after #44107 is now awkward, exactly because of those rename warnings people care so much about.)

In SLNOS we frequently mark patches as "apply no sooner than <date/release>", i.e. in this case of this PR we would do:

- schedule rename + alias in a week (or apply immediately);
- schedule deprecation warning for the next release;
- schedule deletion for the release after the next release.

The beauty of the scheme is that you make all of those in a single go, split them into parts, schedule them in pieces for everyone else, but then always carry and apply the whole patchset in your local branch. This way, if something breaks, you notice it almost immediately and can then rebase and fix the scheduled but not yet applied patches for everyone.

Inb4 "this should go through an RFC process". Though, feel free to make an RFC for this if you feel like it, I won't.

Btw, I propose another related RFC topic: "an official policy on what to do when somebody sneakily reverts (with sneakily defined as: committing without making a PR and/or without mentioning the original author those changes got reverted) something you find useful, but then gets silent after you laid out your arguments against the revert". Not to blame or point fingers at anyone, I agree that the outlined situation is awkward. Hence, lets make an RFC!

In all seriousness, I think Nixpkgs should forbid committing without making a PR and the `git-blame` mention bot.

Infinisil commented:
@oxij I've actually just started writing an RFC for deprecation a few days ago :). And in it I indeed talk about such a two-step deprecation. I will hopefully be able to finish the first draft in a week or so.

I'm also in favor of improving the codebase, removing old code and naming things properly. If we don't allow for that, nixpkgs will have one more factor contributing to its messiness. Even if the time until eventual removal is long, that's still better than keeping old stuff around forever. It's however also important to not break external code, so a smooth deprecation path should be encouraged, which is what my RFC will be about.

Infinisil commented:
And as for my opinion on what to do with this PR: Since we currently don't yet have a set of good functions for deprecation (that would only warn at a later release), I think it's best to keep the old attributes around for compatibility without emitting any warnings. If we keep the warnings, then it might turn out similar to https://github.com/NixOS/nixpkgs/pull/44107, because a consensus wasn't reached.

Also: Having warnings and then removing them again is very bad UX, users could get confused as to why something suddenly got "undeprecated". Not showing any warnings leaves us open with the possibility of adding them later on. Once you show a warning, you can't turn back (ideally).

Ericson2314 commented:
@oxij I was not aware all these renames were being secretly reverted. This is very concerning.

@Infinisil Yes please have the RFC up soon / before the release!! I am willing to help.

Ericson2314 commented:
@volth See the note in the OP. Yes, by analogy that's exactly correct: they should also go. But since they are already a shorthand of sorts already, and since they are so popular, I decided not to fight that just yet.

BTW, the virtually unused longer form is:
```nix
lib.systems.inspect.matchAnyAttrs
  (with lib.systems.inspect.patterns; isi686 // isLinux)
  stdenv.hostPlatform
```
which shows how the (some of the patterns) can be combined. (Some of them are lists, some are sets, sigh.)

dezgeg commented:
Getting rid of top-level `fooPlatform` might be doable given it's fairly recent, even though there clearly appear to be Carnix-generated code that will break if it's removed. But I simply don't understand the desire to go and deliberately cause hassle to people using those names. If you simply change in-tree users of it to the single form of `stdenv.fooPlatform`, most people will copy paste that and get it right. Then most of the rest of the usages will most likely be noticed in code review, where there are (hopefully!) people watching out that code merged to Nixpkgs has at least some form of internal consistency (or if not, you're going to have consistency problems anyway).

The `configureFlags` situation was discussed previously and as far as I can tell the situation with passing arguments with spaces in lists is exactly the same as back then. This is based on @edolstra's comment here:  https://github.com/NixOS/nixpkgs/pull/15799#issuecomment-242747833. So bringing that up again doesn't seem very useful, given that nothing has changed.

> I give you exhibit A, 7ade207, for a rename that never happened, and only got harder.

I am not sure what you mean with that commit reference. Is it that `buildInputs` didn't get renamed to `hostInputs`? That sounds like a good thing to me. The current name is just fine. No other cross compilation distro highlights that the "normal" dependencies as host/target/whatever dependencies either, they are just dependencies. OpenWRT has things like `DEPENDS:=+libuuid`, Yocto has `DEPENDS = "util-linux"`, Buildroot has  `E2FSPROGS_DEPENDENCIES = util-linux`. No mention of 'host' anywhere (though IIRC all of those don't use the autotools convention and call it 'target' instead).

As for deprecating `stdenv.system == "foo"` in favor of `stdenv.hostPlatform.system == "foo"` or `stdenv.isFoo` over `stdenv.hostPlatform.isFoo`, to me each day it feels more and more bad idea, because people who have no practical experience with cross compiling mix up `hostPlatform` and `targetPlatform` constantly (like, literally earlier in this thread). Thus given that `hostPlatform` checks make up like, >95% (yes, I calculated the percentage) of the checks in Nixpkgs, I definitely feel they should be special cased to the shorter forms so that the common form is short compared to the uber rare cases where needs to be used.

Ericson2314 commented:
> This is based on @edolstra's comment here: https://github.com/NixOS/nixpkgs/pull/15799#issuecomment-242747833. So bringing that up again doesn't seem very useful, given that nothing has changed.

Actually, there is now support in Nix 2.0 for sticking the arrays right in bash. Seeing that requiring Nix 2.0+ is something we'll eventually do, we could eventually be in a world without `*Array` where everything indeed works fine. If so, might as well start deprecating now!

dezgeg commented:
Yes, to me top-level `fooPlatform` -> `stdenv.fooPlatform` is a positive change.

Ericson2314 commented:
@dezgeg and the other changes corresponding to the other deprecations?

Ericson2314 commented:
> I am not sure what you mean with that commit reference. Is it that buildInputs didn't get renamed to hostInputs? That sounds like a good thing to me....

I am fine not including "host", but having "build" in the name unfortunate. I would have assumed "buildInput" is the build platform or build-time dependency if I didn't know any better.

But, bikeshedding names aside, my point was that the renames only get harder to so best not to kick the can down the road.

Ericson2314 commented:
> As for deprecating `stdenv.system == "foo"`...

I agree with your huffman encoding, and desire to avoid confusing beginners with unnecessary options. But that's why I'm keeping `stdenv.is*`. Exact `system` checking is *never* good a good idea, whether on `stdenv.hostPlatform.system` or `stdenv.system`. It breaks down with CPU variants (did you really mean `armv6` and not `armv5`?), and also encourages listing all CPUs or all OSes when only one axis matters.

Ericson2314 commented:
> But I simply don't understand the desire to go and deliberately cause hassle to people using those names.

Simply put, they may not even aware they wrote something sub-par.

I defer to what the deprecation RFC looks like, but if we put in a lower bound on the removal date as @oxij says, then we can prioritize depredations and let users choose a threshold at which to care. If these, say, would never be removed, then they would have lowest priority and could even be muted by default.

samueldr commented:
> where there are (hopefully!) people watching out that code merged to Nixpkgs has at least some form of internal consistency (or if not, you're going to have consistency problems anyway).

In the past, I have experiences the anti-pattern where on two different contributions, I was told to go from A to B, then on the second, from B to A. (Don't remember if this was code style, or API use.)

Given a mechanism that keeps the code working, but alerts usage of deprecated (removed **later** in the future / not right now) allows code to keep being backwards compatible, and mechanizes the consistency on those APIs.

<sup>And this comment is out of scope of the technical side of this PR, and should go into considering the deprecation mechanisms and policies later.</sup>

samueldr commented:
> **NOTE**: *This comment has to be read with the express idea that the changes are assumed to be correct and wanted with regards to deprecation and the features, please do not reply derailing this comment if the assumption does not hold. Thank you*

I hate<sup>opinion</sup> how this commit changes too much at the same time. This looks like an omnibus PR, built in a way to hide an agenda into innocuous changes.

What I'm referring to mainly is how disconnected some of the changes are.

My opinion is that this should be split into changes going per topic. Follows my understanding and interpretation

##### Subject 1 [unnamed]

I'm pretty sure this is a unit, and it's all fine to go together. I would probably have made the related changes  in one-commit per change just because, but then this becomes noisier in commits for no real value (so a +1-1 on that last sentence).

 * `system ==> stdenv.buildPlatform.system`
 * `stdenv.system ==> stdenv.buildPlatform.system`
 * `buildPlatform ==> stdenv.buildPlatform`
 * `hostPlatform ==> stdenv.hostPlatform`
 * `targetPlatform ==> stdenv.targetPlatform`

##### Subject 2: Improving an existing deprecation message

(If the improved message requires the changes from subject 1, it should be opened in relation to that change.)

 * `isArm ==> isAarch32 # Already done during 18.03 but message improved`

##### Subject 3: Deprecate use of string `configureFlags`

This one AFAIUI does not relate at all to the two previous subject. This is another PR. This should be handled separately, with a different technical discussion.

 * `configureFlags = "--foo --bar"; ==> configureFlags = [ "--foo" "--bar" ];`


 * * * 

> PRs are not threaded, discussing disjointed changes, or having parallel discussion can quickly become arduous. Splitting the PR would at least keep the technical side of the discussion easier to follow.

Ericson2314 commented:
Yes it started just a 1, but then I discovered 2 and 3 in the process. I'll submit new changes for the churn, new message, and `configureFlags` deprecation, leaving this one for the original deprecation.

Ericson2314 commented:
OK, https://github.com/NixOS/nixpkgs/pull/45820 just changes the existing usages within nixpkgs without formally deprecating anything.

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=7cbfd008-327d-4576-ac5b-a35e345c7882)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=5dcf8ae0-c246-4745-b61b-1c639f1d3c4f)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=e5630c96-95a3-4cc0-8afd-cb3387600b0c)

The following builds were skipped because they don't evaluate on x86_64-linux: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `stdenv' because:
GC Warning: Failed to expand heap by 10737418240 bytes
Can't start up: not enough memory

```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=92e8f8c4-5bac-4fae-a345-5f7916a3e0fc)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=4d9c864b-bed0-4b66-985d-8dd288ac6c5e)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=aa821620-0f71-4a04-9fbc-fc45316b592a)

The following builds were skipped because they don't evaluate on x86_64-linux: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `stdenv' because:
GC Warning: Failed to expand heap by 10737418240 bytes
Can't start up: not enough memory

```
</p></details>

Ericson2314 commented:
OK this is just the platform deprecations, and the thread of removal after 18.09 is no longer mentioned.

GrahamcOfBorg commented:
<!--REQUEST_ID=e461296c-6f93-4c9d-86b4-a49b6da8c2f2-->Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=418cb973-183a-4853-826f-f6d224fcde19)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=6c7b5aa8-1cac-4b0a-8d27-11c448ccb3bc-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=06102a61-0569-43e5-8349-caa575b3985e)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=6c7b5aa8-1cac-4b0a-8d27-11c448ccb3bc-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=da426e39-b2db-4550-9f15-af8f51b7753f)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e461296c-6f93-4c9d-86b4-a49b6da8c2f2-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=e30c5d1f-7067-4e2c-aef4-47460adf224a)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/i6vl5lwlz5jbkg4r6p340dwmj6fha3xq-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=6c7b5aa8-1cac-4b0a-8d27-11c448ccb3bc-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=111abacf-61cb-44df-8ba8-73477d9edd53)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/i6vl5lwlz5jbkg4r6p340dwmj6fha3xq-stdenv-linux
```
</p></details>

Ericson2314 commented:
configure flags bit moved to https://github.com/NixOS/nixpkgs/pull/45886.

dezgeg commented:
> In the past, I have experiences the anti-pattern where on two different contributions, I was told to go from A to B, then on the second, from B to A. (Don't remember if this was code style, or API use.)

Yes I can see how that can happen for certain style/convention discussions (I think I have seen the exact thing happen when discussing whether `version = "1.2.3";` belongs in a `let` block or the derivation itself). But I don't buy it for this case of `stdenv.fooPlatform` vs` fooPlatform` since a one or two `git grep`s immediately gives away the answer to which is the more conventional one.

dezgeg commented:
> I now surmise that many people here simply hate the warning messages.

Yes, if there is a warning which I cannot disable, it might just as well have been an fatal error. Otherwise it just drowns out the real errors/warnings in the noise.

timokau commented:
> Yes, if there is a warning which I cannot disable, it might just as well have been an fatal error. Otherwise it just drowns out the real errors/warnings in the noise.

If we fix all uses in nixpkgs it would not be noise. It would alert out of tree users to the change without breaking their workflow immediately. I think that is the exact right use of a warning.

Ericson2314 commented:
> But you do need to check for the specific architecture from time to time.

That is not OS-specific, so it should be done not with `system` but with:
```nix
stdenv.hostPlatform.isArm && stdenv.hostPlatform.parsed.cpu.version <= 6
```
That's a totally legit use-case, but not one which I think is common enough to deserve a shorthand.

> I don't see how that's related at all...

That was just an off-topic aside, but I meant that not caring about the build platform is a good sign that cross builds ought to work. If a package does care about the build platform, then they need also *explicitly* allow varying the host and build platforms independently for cross to work. But if it truly doesn't care, then they are allowed to vary independently by parametricity.

Ericson2314 commented:
> For many of the cases, people cannot have possibly written...

Exactly! When they wrote the code, it wasn't sub par. And if they either haven't followed the cross stuff, or forget about the code they wrote---both highly likely scenarios on their own---they'll have no idea the code can be improved.

> Warning which I cannot disable

I have cleaned up Nixpkgs so all instances of the warning will be from user code they *can* fix. And I have no problem making deprecations disabled with some `config` option.

Ericson2314 commented:
@Infinisil pkgs/applications/editors/neovim/wrapper.nix that is the once extent deprecation I found so far, for sake of your RFC or a PR showing what it would look like in practice.

matthewbauer commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -127,6 +127,11 @@ let
         "`stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead"
         hostPlatform.isAarch32;
 
+      system = builtins.trace
```
I think we want to use lib.warn here where possible.

Ericson2314 replied:
Thanks! Fixed in https://github.com/NixOS/nixpkgs/pull/46059

matthewbauer commented on pkgs/top-level/stage.nix:
```diff
@@ -79,11 +79,20 @@ let
-    inherit buildPlatform hostPlatform targetPlatform;
-    inherit (buildPlatform) system;
+  platformCompat = self: super: {
+    buildPlatform = builtins.trace
```
I'm not sure if we need to deprecate these. To me it seems like the difference between using ```lib`` or ```stdenv.lib```. No real benefits for one over the other.

Ericson2314 replied:
I agree these deprecations, unlike the others, are "fully explicit" so there's less a downside. But still, I think it benefits no one to have redundant bindings---I'd actually be for deprecating one of `lib` or `stdenv.lib` too for the same reason.

edolstra commented:
Don't remove `system` and `stdenv.system`. I can remember those but not `stdenv.buildPlatform.system`. Also, how are these better, and who declared them to be obsolete? Even #45820 seems premature to me.

On deprecation warnings or removals: they're very annoying in out-of-tree modules that have to support multiple Nixpkgs versions.

edolstra commented:
To expand a bit: The current cross-compilation infrastructure was merged on the premise that it wouldn't affect users who don't care about cross-compilation, but being forced to write `stdenv.buildPlatform.system` *does* affect them.

Ericson2314 commented:
> To expand a bit: The current cross-compilation infrastructure was merged on the premise that it wouldn't affect users who don't care about cross-compilation, but being forced to write stdenv.buildPlatform.system does affect them.

Yes I agree this is the heart of the matter. By proposing deprecating these things, I'm beginning to end the experimental phase of cross compilation experiment with the conclusion that it was successful. My intent is *is* to force people to give 5 seconds worth of thought to cross compilation if the `stdenv.is*` predicates don't work for their use-cases (and they do support the *vast* majority of use-cases). I don't think that is too onerous at all.

> I can remember those but not `stdenv.buildPlatform.system`.

Have you *tried* using them? Maybe with my deprecation notice you would remember them.

> how are these better?

`system` is rarely a good way to do anything but choose prebuilt binaries. (Even then, it is bad for platforms with more than one ABI and `hostPlatform.config` is a better choice). I don't want to promote the attribute which is poorly suited to most tasks. I rather push people towards `stdenv.is*`.

> who declared them to be obsolete

This PR is proposing that they be so declared. If that were already the consensus I wouldn't have gone asking for review in multiple places.

> On deprecation warnings or removals: they're very annoying in out-of-tree modules that have to support multiple Nixpkgs versions.

Is this a real use-case? Every commercial project I've seen pins an exact nixpkgs revision for stability. Using old stable channels is insecure and not recommended.

Regardless, all of these alternatives date back to at least 17.09. I think 3 stable releases (17.09, 18.03, 18.09) is plenty.

matthewbauer commented:
I think the main reason to deprecate stdenv.system is the confusion as to what system it is applying to. Is it the system building the software or the system that will run the software. This applies to ```stdenv.is*``` as well, though. I think usually when you look at system you are wanting the one that the software is built _for_ not what it was built _on_ (obviously many exceptions though).

Ericson2314 committed:
stage: Deprecate `system` for `stdenv.system`

I have my own reasons, but these should be most palatable to others:

 - @Dezgeg prefers `stdenv.*Platform` to `*Platform`, so by analogy it's
 always better to refer to properties of the standard environment from
 the standard environment.

 - @edolstra is trying to limit the presence of non-packages at the top
 level. `system` is not a derivation.

Ericson2314 commented:
OK in #46059 #46146 I deprecated top-level `buildPlatform`, `hostPlatform`, and `targetPlatform`, since they are newer, part of the cross experiment, and of my own creation. With the previous creation of the `configureFlags` PR, that just leaves `stdenv.system` and  `system` in this PR.
@edolstra and @dezgeg Say we just deprecate top-level `system`, which is now the first commit (80519362636c7ac08d3f3344ec6c62149bd248e5). This keeps what you both see as the benefits of 1) not thinking about cross 2) good host platform default 3) it's quite old. @dezgeg this is also in keeping with promoting `stdenv.*Platform` as the better option: let information about the stdenv come from stdenv. @edolstra you have previously expressed an interest in keeping non-derivations (or even non "packages"; things that people might want to `nix run`) out of the top level. This is also in keeping of that.

If that sounds alright, I'll make that

GrahamcOfBorg commented:
<!--REQUEST_ID=51b29698-e548-40a0-ae8f-56d8b8de7b16-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=92a94c95-3265-4aa3-b0d1-6388888ff924)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/95bw3cdcgc8c0rrvqwvhvc805pdv7rmx-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=51b29698-e548-40a0-ae8f-56d8b8de7b16-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45717&attempt_id=08d80149-b982-4abd-82b4-16c4d8d2cc93)

Attempted: stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pjq5v1xgljr4ygb1h1xfybdjydczcqiw-stdenv-linux
```
</p></details>

oxij commented:
Current state of this PR's changes is confusing:
> "... Please use `stdenv.system` or `stdenv.hostPlatform.system`."
> "`stdenv.system` is deprecated since 18.09. Please use `stdenv.hostPlatform.system`."

Ericson2314 commented:
Hehe that's me hoping the first PR is more palatable than the second. It would be nice to at least get the whole suite of "top-level depreciations" if they both really would like `stdenv.system` to stay in place.
.system`.

edolstra commented:
> Is this a real use-case? 

Yes, this is a real use-case: NixOps has to support multiple NixOS versions. So deprecation warnings are annoying because either we get warnings on newer NixOS versions, or we fix the warnings and break support for the older versions.

Ericson2314 commented:
What versions of NixOS does NixOPS currently attempt to support? Is this documented anywhere?

Ericson2314 commented:
Also, I offer to personally backport `stdenv.*Platform.system` to prior NixOS versions.

Ericson2314 commented:
> These system strings are what I and other people know, this other stuff I have no idea even existed.

So what? The deprecation warning brings awareness to the new interface. It's absurd to argue against advertising the new interface and then complain it's not well known.

> Will it return "5tel" or 5 on armv5tel-linux?

5

> e.g what's this stdenv.hostPlatform.parsed.cpu.version even gonna return on an x86_64?

The field is only present for ARM, as it's not clear what it should be elsewhere when we don't have multiple versions anyways.

> Now yes that will assume Linux, but that's the only OS supported on ARMv6 anyway

iOS. I do use Nix and nixpkgs to build iOS apps for work, so this is not at all hypothetical.

But that to me is besides the point. Except for things like prebuilt binaries, it's usually clear when some change is because of the OS or because of the CPU. If say, we're supporting `aarch64-linux`, some `arm-linux`, and `x86_64-linux`, two orthogonal conditions is much better style than one fat condition on the system. It's crucial to push the interfaces coax developers in the direction of better code.

> In fact I don't even know if -latomic is needed on non-GCC-using platforms or is it actually going to break things there.

It is not unlikely one would want to use Clang for these things. compiler-rt provides atomics too, but either library can be used.

> Same thing, if there is no real practical difference, the interface that people know is better.

As an aside, I'm not even sure it is a better known interface. Outside of Nix I probably see `x86_64-unknown-linux-gnu` more than `x86_64-linux`.

But yes, I do concede at the moment prebuilt binaries tend do only bother with one ABI per CPU+OS.

edolstra commented:
IMHO, we should really reconsider whether these gratuitous deprecations/removals/renames are a good idea. They all add a little bit of friction to each NixOS upgrade, which, given enough of them, adds up to a lot of friction. Ideally you should be able to try out a newer NixOS version without having to make lots of little changes to your configuration modules / Nix packages, and to use those in multiple NixOS versions without having to put in lots of version conditionals.

flokli commented:
I guess this won't make it into 18.09 or 19.03 ;-)

I updated the PR title. I think we should come to a consensus about deprecations. Could somebody revive [RFC 0033](https://github.com/NixOS/rfcs/pull/33)?

Ericson2314 commented:
Yes this is blocked on that RFC, which I do mean to get back to. (I am technically coauthor but haven't done anything yet!! :astonished:).

Mic92 commented:
rfc33 was closed. Should we keep this one open?

flokli commented:
As stated by @Ericson2314 in https://github.com/NixOS/nixpkgs/pull/45717#issuecomment-478685802, this is blocked on an RFC, which is yet to be revived.

I propose closing it for now - as the PR doesn't apply cleanly either and is not actionable atm.

Ericson2314 commented:
Find with me
